crates/accelerate/src/euler_one_qubit_decomposer.rs

```diff
@@ -853,6 +965,110 @@ pub fn params_zxz(unitary: PyReadonlyArray2<Complex64>) -> [f64; 4] {
     params_zxz_inner(mat)
 }
 
+type OptimizeDecompositionReturn = Option<((f64, usize), (f64, usize), OneQubitGateSequence)>;
+
+#[pyfunction]
+pub fn optimize_1q_gates_decomposition(
+    runs: Vec<Vec<PyRef<DAGOpNode>>>,
+    qubits: Vec<usize>,
+    bases: Vec<Vec<PyBackedStr>>,
+    simplify: bool,
+    error_map: Option<&OneQubitGateErrorMap>,
+    atol: Option<f64>,
+) -> Vec<OptimizeDecompositionReturn> {
+    runs.iter()
+        .enumerate()
+        .map(|(index, raw_run)| -> OptimizeDecompositionReturn {
+            let mut error = match error_map {
+                Some(_) => 1.,
+                None => raw_run.len() as f64,
+            };
+            let qubit = qubits[index];
+            let operator = &raw_run
+                .iter()
+                .map(|node| {
+                    if let Some(err_map) = error_map {
+                        error *=
+                            compute_error_term(node.instruction.operation.name(), err_map, qubit)
+                    }
+                    node.instruction
+                        .operation
+                        .matrix(&node.instruction.params)
+                        .unwrap()
+                })
+                .fold(
+                    [
+                        [Complex64::new(1., 0.), Complex64::new(0., 0.)],
+                        [Complex64::new(0., 0.), Complex64::new(1., 0.)],
+                    ],
+                    |mut operator, node| {
+                        matmul_1q(&mut operator, node);
+                        operator
+                    },
+                );
+            let old_error = if error_map.is_some() {
+                (1. - error, raw_run.len())
+            } else {
+                (error, raw_run.len())
+            };
+            let target_basis_vec: Vec<EulerBasis> = bases[index]
+                .iter()
+                .map(|basis| EulerBasis::__new__(basis).unwrap())
+                .collect();
+            unitary_to_gate_sequence_inner(
+                aview2(operator),
+                &target_basis_vec,
+                qubit,
+                error_map,
+                simplify,
+                atol,
+            )
+            .map(|out_seq| {
+                let new_error = compute_error_one_qubit_sequence(&out_seq, qubit, error_map);
+                (old_error, new_error, out_seq)
+            })
+        })
+        .collect()
+}
+
+fn matmul_1q(operator: &mut [[Complex64; 2]; 2], other: Array2<Complex64>) {
```

Comment:
I'm not sure where this would save an allocation this is just destructuring the array reference to individual variables. I would expect the compiler to generate the same exact code for this as the code in this function already.

The one allocation point I'm not happy about but I couldn't really figure out how to avoid was around the `Array2` for `other` which we're getting from `Operation::matrix()`. I needed to do that from a typing perspective in the `Operation` trait, because using an array view wouldn't work since we may get the array view from Python or it might point to a static array so I couldn't see a way to declare a lifetime parameter for the trait method that would work (or at least I think that's why it returns an owned array, it's been >1 month since I wrote that class so I'm not 100%).

Reply:
Another thing. We can test if all gates are standard, and if so do some optimizations. Like use fixed sized `Array`s throughout the multiplication. Doesn't need to be in this PR of course.

Reply:
Yeah, let's look at that kind of thing in a followup. Right now the runtime overhead is mostly dominated by the rust python boundary and I think we'll have another round or two of optimizations to perform on this pass after #12550 merges. We also should have more flexibility when everything is handled in rust.

Reply:
> just destructuring the array reference to individual variables. 

I see that now. In fact, if there were an allocation, there is no way the caller know where it is.

Reply:
> avoid was around the `Array2` for other which we're getting from `Operation::matrix()`

This is the optimization I was talking about above. I didn't think it through all the way. But if you could scan or use a flag to know that all matrices are slices of fixed size, then multiplication would be much faster. They are all 2x2 of course, but `Array2` doesnt carry the info in the type.